Exit early if domain file is invalid

joaorobson opened the pull request:
**Proposed changes**:

Closes #4619
* The command `rasa data validate` must exit early if the domain can't be loaded, avoiding a bunch of errors to be shown.

**Expected behavior**:

* If the domain file is configurated as shown below:
```
intents:
  - greet

actions:
- utter_greet
- utter_greet

templates:
  utter_greet:
  - text: "Hey! How are you?"
```
the log should be similar to this:

```
2019-10-23 12:11:26 WARNING  rasa.importers.rasa  - Loading domain from 'domain.yml' failed. Using empty domain. Error: 'Duplicate actions in domain. These actions occur more than once in the domain: 'utter_greet''
2019-10-23 12:11:26 WARNING  rasa.importers.rasa  - Loading domain from 'domain.yml' failed. Using empty domain. Error: 'Duplicate actions in domain. These actions occur more than once in the domain: 'utter_greet''
2019-10-23 12:11:26 WARNING  rasa.core.training.dsl  - Found unknown intent 'greet' on line 2. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:11:26 WARNING  rasa.core.training.dsl  - Found unknown intent 'mood_great' on line 4. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:11:26 WARNING  rasa.core.training.dsl  - Found unknown intent 'greet' on line 8. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:11:26 WARNING  rasa.core.training.dsl  - Found unknown intent 'mood_unhappy' on line 10. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:11:26 WARNING  rasa.core.training.dsl  - Found unknown intent 'affirm' on line 13. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:11:26 WARNING  rasa.core.training.dsl  - Found unknown intent 'greet' on line 17. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:11:26 WARNING  rasa.core.training.dsl  - Found unknown intent 'mood_unhappy' on line 19. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:11:26 WARNING  rasa.core.training.dsl  - Found unknown intent 'deny' on line 22. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:11:26 WARNING  rasa.core.training.dsl  - Found unknown intent 'goodbye' on line 26. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:11:26 WARNING  rasa.core.training.dsl  - Found unknown intent 'bot_challenge' on line 30. Please, make sure that all intents are listed in your domain yaml.
```
 and not this:

```
2019-10-23 12:12:04 WARNING  rasa.importers.rasa  - Loading domain from 'domain.yml' failed. Using empty domain. Error: 'Duplicate actions in domain. These actions occur more than once in the domain: 'utter_greet''
2019-10-23 12:12:04 WARNING  rasa.importers.rasa  - Loading domain from 'domain.yml' failed. Using empty domain. Error: 'Duplicate actions in domain. These actions occur more than once in the domain: 'utter_greet''
2019-10-23 12:12:04 WARNING  rasa.core.training.dsl  - Found unknown intent 'greet' on line 2. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:12:04 WARNING  rasa.core.training.dsl  - Found unknown intent 'mood_great' on line 4. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:12:04 WARNING  rasa.core.training.dsl  - Found unknown intent 'greet' on line 8. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:12:04 WARNING  rasa.core.training.dsl  - Found unknown intent 'mood_unhappy' on line 10. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:12:04 WARNING  rasa.core.training.dsl  - Found unknown intent 'affirm' on line 13. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:12:04 WARNING  rasa.core.training.dsl  - Found unknown intent 'greet' on line 17. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:12:04 WARNING  rasa.core.training.dsl  - Found unknown intent 'mood_unhappy' on line 19. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:12:04 WARNING  rasa.core.training.dsl  - Found unknown intent 'deny' on line 22. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:12:04 WARNING  rasa.core.training.dsl  - Found unknown intent 'goodbye' on line 26. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:12:04 WARNING  rasa.core.training.dsl  - Found unknown intent 'bot_challenge' on line 30. Please, make sure that all intents are listed in your domain yaml.
2019-10-23 12:12:04 INFO     rasa.core.validator  - Validating intents...
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'goodbye' is in the NLU training data, but is not listed in the domain.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'affirm' is in the NLU training data, but is not listed in the domain.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'mood_great' is in the NLU training data, but is not listed in the domain.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'bot_challenge' is in the NLU training data, but is not listed in the domain.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'deny' is in the NLU training data, but is not listed in the domain.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'greet' is in the NLU training data, but is not listed in the domain.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'mood_unhappy' is in the NLU training data, but is not listed in the domain.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'goodbye' is used in stories, but is not listed in the domain file.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'affirm' is used in stories, but is not listed in the domain file.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'mood_great' is used in stories, but is not listed in the domain file.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'bot_challenge' is used in stories, but is not listed in the domain file.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'deny' is used in stories, but is not listed in the domain file.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'greet' is used in stories, but is not listed in the domain file.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The intent 'mood_unhappy' is used in stories, but is not listed in the domain file.
2019-10-23 12:12:04 INFO     rasa.core.validator  - Validating there is no duplications...
2019-10-23 12:12:04 INFO     rasa.core.validator  - Validating utterances...
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The utterance 'utter_greet' is used in stories, but is not a valid utterance.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The utterance 'utter_happy' is used in stories, but is not a valid utterance.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The utterance 'utter_cheer_up' is used in stories, but is not a valid utterance.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The utterance 'utter_did_that_help' is used in stories, but is not a valid utterance.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The utterance 'utter_goodbye' is used in stories, but is not a valid utterance.
2019-10-23 12:12:04 ERROR    rasa.core.validator  - The utterance 'utter_iamabot' is used in stories, but is not a valid utterance.
```



**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RasaHQ/rasa?pullRequest=4661) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/RasaHQ/rasa?pullRequest=4661) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: ViniciusBernardo<br/>:x: lucianaribeiro<br/>:x: joaorobson<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RasaHQ/rasa?pullRequest=4661) it.</sub>

erohmensing commented:
@joaorobson thanks for the PR. Is this ready for review? If so can you mark it as so? :)

joaorobson commented:
> @joaorobson thanks for the PR. Is this ready for review? If so can you mark it as so? :)

@erohmensing, It was lacking a unit test that I just added. If the build runs successfully, the PR will be ready for review.

erohmensing commented on rasa/core/validator.py:
```diff
@@ -199,3 +199,14 @@ def verify_all(self, ignore_warnings: bool = True) -> bool:
+        """ Runs the validations on domain."""
+
+        is_valid = not self.verify_domain_is_empty()
+        return is_valid
```
To be clear, the way this is working is that if the domain is invalid, the importer won't return anything, therefore `self.domain.is_empty()` is true?

If so, I think it makes more sense to merge these 2 functions(one just calls the other anyway) into one called e.g. `verify_domain_validity`

joaorobson replied:
Yeah, it makes total sense!

erohmensing commented on rasa/cli/data.py:
```diff
@@ -103,5 +103,9 @@ def validate_files(args):
 
     validator = loop.run_until_complete(Validator.from_importer(file_importer))
+    domain_is_alright = validator.verify_domain()
+    if not domain_is_alright:
```
```suggestion
    if not domain_is_valid:
```

erohmensing commented on rasa/cli/data.py:
```diff
@@ -103,5 +103,9 @@ def validate_files(args):
     )
 
     validator = loop.run_until_complete(Validator.from_importer(file_importer))
+    domain_is_alright = validator.verify_domain()
```
```suggestion
    domain_is_valid = validator.verify_domain()
```

erohmensing commented on docs/user-guide/validate-files.rst:
```diff
@@ -41,6 +41,10 @@ actions.
 
+**verify_domain_is_empty():** Check if domain returned by the importer is empty.
+
+**verify_domain():** Check if domain is valid.
```
```suggestion
**verify_domain_validity():** Check if domain is valid.
```

erohmensing commented on rasa/core/validator.py:
```diff
@@ -199,3 +199,8 @@ def verify_all(self, ignore_warnings: bool = True) -> bool:
         return intents_are_valid and stories_are_valid and there_is_no_duplication
+
+    def verify_domain_validity(self) -> bool:
+        """ Runs the validations on domain."""
```
```suggestion
        """Checks whether the domain returned by the importer is empty, indicating an invalid domain."""
```

erohmensing commented on tests/cli/test_rasa_data.py:
```diff
@@ -1,6 +1,9 @@
+from collections import namedtuple
 from typing import Callable
 from _pytest.pytester import RunResult
+from rasa.cli.data import validate_files
```
```suggestion
from rasa.cli import data
```
We try to import modules rather than methods as much as possible, to make it clear where methods are coming from

joaorobson replied:
> Functionality looks great  Just some style comments and a revision to the docs after what we changed last time and we're good.
> 
> P.S. sorry to have you waiting and re-merging master, I wasn't aware you were ready (didn't get a notification), in the future feel free to just drop a comment when all the mentioned changes have been integrated and it's ready for another look!

Ok, no problem : )

erohmensing commented on tests/cli/test_rasa_data.py:
```diff
@@ -63,3 +66,12 @@ def test_data_validate_help(run: Callable[..., RunResult]):
+def test_validate_files_exit_early():
+    with pytest.raises(SystemExit) as pytest_e:
+        args = {"domain": "data/test_domains/duplicate_intents.yml", "data": None}
+        validate_files(namedtuple("Args", args.keys())(*args.values()))
```
```suggestion
        data.validate_files(namedtuple("Args", args.keys())(*args.values()))
```

erohmensing commented on tests/core/test_validator.py:
```diff
@@ -88,6 +89,29 @@
+        "Loading domain from '{}' failed. Using empty domain. "
+        "Error: 'Intents are not unique! Found two intents with name "
+        "'default'. Either rename or remove one of them.'".format(
+            domain_path, full_path
```
Looks like there is only one argument being passed to this `.format()`. Also we're moving to f-strings, that just hasn't been merged to master yet, would you use one here as well?

joaorobson replied:
> Functionality looks great  Just some style comments and a revision to the docs after what we changed last time and we're good.
> 
> P.S. sorry to have you waiting and re-merging master, I wasn't aware you were ready (didn't get a notification), in the future feel free to just drop a comment when all the mentioned changes have been integrated and it's ready for another look!

Ok, no problem : )

erohmensing commented on tests/core/test_validator.py:
```diff
@@ -88,6 +89,29 @@
     )
 
 
+async def test_verify_domain_validity(caplog):
```
```suggestion
async def test_early_exit_on_invalid_domain(caplog):
```

erohmensing commented on docs/user-guide/validate-files.rst:
```diff
@@ -41,6 +41,10 @@ actions.
 **verify_all():** Runs all verifications above.
 
+**verify_domain_is_empty():** Check if domain returned by the importer is empty.
+
```
Remove 44 and 45 after your we updated this

joaorobson commented:
> Functionality looks great  Just some style comments and a revision to the docs after what we changed last time and we're good.
> 
> P.S. sorry to have you waiting and re-merging master, I wasn't aware you were ready (didn't get a notification), in the future feel free to just drop a comment when all the mentioned changes have been integrated and it's ready for another look!

@erohmensing I guess that I've fixed everything you requested.
